### PR TITLE
converted to TS

### PR DESCRIPTION
**Pull request by FunkyDunk-Z**
Alrighty so here we go. Most of the codebase has bee converted to TS. 
I haven't started on Characters just yet but i want to get this part going so we can begin to incorporate the turn tracker already.

A few key points are: 

**useCustomFetch Hook** 
I renamed this to avoid confusion with fetch. customFetch takes in a number of options with the only optional one being dataToSend as we dont always need this. we could set credentials to be default of true and set them to false when we need but for now we have to set it each time. along with the url, the fetchType and authType.

**isLoggedIn**
Currently i am checking this in App.tsx, we could move it to AuthContext but i was getting some double renders and this was the solution for the time being.

**Backend - crudOps**
so far i only have createOne set up as first i want to get the character model and controller set up and see if its worth having a generic crudOps or if each model will do something slightly different.

**Forgot Password Page**
Not set up yet.

There's quite a bit of code to go through so I probably missed a bunch of important stuff, but so far everything is working

### CODE REVIEW

**Comment by FunkyDunk-Z** [comment]
Also Error handling is almost non exsitant right now, sorry for that !

**Comment by FunkyDunk-Z** [comment]
Another important info is that the api url for the backend is now /api/v2/... it is no longer v1

File: client/.env [hunk 1]
```diff
@@ -0,0 +1,2 @@
+VITE_EMAIL=dunc@gmail.com
```
Comment (luke-dowling) [inline client/.env]: dont push env files togithub- unless it is an example of how to use? otherwise remove this file

File: client/.gitignore [hunk 2]
```diff
@@ -22,4 +22,3 @@ dist-ssr
 *.njsproj
 *.sln
 *.sw?
```
Comment (luke-dowling) [inline client/.gitignore:24]: please reset and add .env file

File: client/.prettierrc
Comment (luke-dowling) [inline client/.prettierrc:1]: why did you delete prettier?

**Review by luke-dowling — CHANGES_REQUESTED** [review]
please remove the node modules from the commit

**Comment by FunkyDunk-Z** [comment]
alrighty, so that should be node_modules and .env removed and prettier added on the client side

File: client/.gitignore
Comment (luke-dowling) [inline client/.gitignore:1]: please also add the dist folder to the .gitignore

File: client/src/components/layout/Header.tsx [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+  const handleNavigate = () => {
+    navigate("/");
+  };
+
+  return (
+    <div id="header" className={styles.wrapper}>
+      <h1 className={styles.title} onClick={handleNavigate}>
+        Mythcarver
+      </h1>
+      {user ? <Navbar /> : ""}
```
Comment (luke-dowling) [inline client/src/components/layout/Header.tsx]: can we use null instead of empty strings for optional rendering, because an empty string is still a data where something is there but null will just not show anything -> both will work but I think null makes more sense?

File: client/src/components/layout/Navbar.tsx [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+// css
+import styles from "./css/Navbar.module.css";
+
+function Navbar() {
+  const { user } = useAuthContext();
+  const { customFetch } = useCustomFetch();
+  const mobile = window.innerWidth < 900;
+  const [isMobile, setIsMobile] = useState(mobile);
+  const [isActive, setIsActive] = useState(false);
+  const [isOpen, setIsOpen] = useState("");
```
Comment (luke-dowling) [inline client/src/components/layout/Navbar.tsx]: can we change the name of isOpen to openStatus (or some name like that) -> isOpen makes it seem like its gonna be a boolean

File: client/src/components/layout/css/Header.module.css [hunk 5]
```diff
@@ -1,45 +1,28 @@
-@import '../../../Global.css';
+@import "../../../Global.css";
```
Comment (luke-dowling) [inline client/src/components/layout/css/Header.module.css]: why are you importing global css to the modules? is this a variables thing?

File: client/src/data/Categories.ts [hunk 6]
```diff
@@ -0,0 +1,38 @@
+const Categories = [
```
Comment (luke-dowling) [inline client/src/data/Categories.ts:1]: can we have this as a list of strings instead of objects because the all have the same keys?

File: client/src/pages/Dashboard.tsx [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import Image from "../assets/Campaigns.webp";
+
+import styles from "./css/Dashboard.module.css";
+
+function Dashboard() {
+  const { user } = useAuthContext();
+
+  return (
+    <div className={styles.wrapper}>
+      <>
```
Comment (luke-dowling) [inline client/src/pages/Dashboard.tsx]: delete fragment

File: client/src/vite-env.d.ts
Comment (luke-dowling) [inline client/src/vite-env.d.ts:1]: we could use this file and put all the shared types in here, that would mean we dont have to reference the shared types file whenever we want to use a type in that file -> this is like global types and we dont have to reference the file we can just use them anywhere?

File: server/dist/app.js
Comment (luke-dowling) [inline server/dist/app.js:1]: please add dist folder to git ignore

File: server/src/app.ts [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+const app = express()
+
+if (env.isDevelopment) {
+  app.use(morgan('dev'))
+}
+
+app.use(
+  cors({
+    origin: ['http://localhost:3000', 'http://localhost:5173'],
+    methods: ['GET', 'POST', 'PUT', 'DELETE'],
```
Comment (luke-dowling) [inline server/src/app.ts]: patch not put

File: server/src/models/characterModel.ts [hunk 9]
```diff
@@ -174,15 +248,15 @@ const characterSchema = new Schema({
       },
     },
   },
-  speeds: {
+  speed: {
     walking: {
       type: Number,
       default: 35,
```
Comment (luke-dowling) [inline server/src/models/characterModel.ts]: default 30?

File: server/src/models/characterModel.ts [hunk 10]
```diff
@@ -270,6 +350,21 @@ characterSchema.pre('save', function (next) {
   next()
 })
 
-const Character = mongoose.model('Character', characterSchema)
+characterSchema.pre('save', async function (next) {
+  const user = await User.findById(this.createdBy)
```
Comment (luke-dowling) [inline server/src/models/characterModel.ts]: try catch block in case the codex goes wrong?

File: server/src/routes/characterRoutes.ts [hunk 11]
```diff
@@ -0,0 +1,12 @@
+import { Router } from 'express'
+
+import { protect } from '../controllers/authController'
+import { createCharacter } from '../controllers/characterController'
+
+const router = Router()
+
+router.use('/', protect)
+
+router.post('/create-character', createCharacter)
```
Comment (luke-dowling) [inline server/src/routes/characterRoutes.ts:14]:
```suggestion
router.route("/create-character).post(protect, createCharacter)
```

File: server/src/routes/characterRoutes.ts [hunk 12]
```diff
@@ -0,0 +1,12 @@
+import { Router } from 'express'
+
+import { protect } from '../controllers/authController'
+import { createCharacter } from '../controllers/characterController'
+
+const router = Router()
+
+router.use('/', protect)
```
Comment (luke-dowling) [inline server/src/routes/characterRoutes.ts:11]:
```suggestion

```

File: server/src/util/validateEnv.ts [hunk 13]
```diff
@@ -0,0 +1,21 @@
+import { cleanEnv } from 'envalid';
+import { port, str } from 'envalid/dist/validators';
+
+export default cleanEnv(process.env, {
+  PORT: port(),
```
Comment (luke-dowling) [inline server/src/util/validateEnv.ts]:
```suggestion
  PORT: port() || 8080,
```

File: server/src/util/validateEnv.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+import { cleanEnv } from 'envalid';
+import { port, str } from 'envalid/dist/validators';
```
Comment (luke-dowling) [inline server/src/util/validateEnv.ts]: ive not used this package so I will have to look up what it does but cool

**Review by luke-dowling — CHANGES_REQUESTED** [review]
Couple of changes in there but looks great, take a look through the comments and let me know if theres any questions

File: client/src/components/layout/css/Header.module.css (hunk 5 above)
Comment (FunkyDunk-Z) [reply client/src/components/layout/css/Header.module.css]: yes it was originally for the colour variables but right now im not using them so i have taken out the import. we can add it back in if we want to use those variables.

File: client/src/components/layout/Header.tsx (hunk 3 above)
Comment (FunkyDunk-Z) [reply client/src/components/layout/Header.tsx]: yep totally makes sense, changed

File: client/src/components/layout/Navbar.tsx (hunk 4 above)
Comment (FunkyDunk-Z) [reply client/src/components/layout/Navbar.tsx]: done

File: client/src/data/Categories.ts (hunk 6 above)
Comment (FunkyDunk-Z) [reply client/src/data/Categories.ts:1]: done and is less code, nice !

File: server/src/models/characterModel.ts (hunk 10 above)
Comment (FunkyDunk-Z) [reply server/src/models/characterModel.ts]: if you see anymore try catch blocks that need to be there let me know, i never really know when its good to use. or should i just always put one in?

File: server/src/routes/characterRoutes.ts (hunk 11 above)
Comment (FunkyDunk-Z) [reply server/src/routes/characterRoutes.ts:14]: So right now i have it set like this as there will be more routes, an update, delete and get, so instead of writing (protect, ...) on each one i wanted to call it once before all the routes. at least thats what Jonas did on his course, do you think it would be better to call protect on each route?

File: server/src/util/validateEnv.ts (hunk 14 above)
Comment (FunkyDunk-Z) [reply server/src/util/validateEnv.ts]: yer i also dont really have much knowledge but it seems to work for now

File: client/src/vite-env.d.ts
Comment (FunkyDunk-Z) [reply client/src/vite-env.d.ts:1]: Nice ! this saves a lot of time and effort writing the imports !

File: server/src/app.ts (hunk 8 above)
Comment (FunkyDunk-Z) [reply server/src/app.ts]: that would explain why my update routes werent working correctly !!!

File: server/src/models/characterModel.ts (hunk 9 above)
Comment (FunkyDunk-Z) [reply server/src/models/characterModel.ts]: that was me just testing out the math function, its 30 now

File: server/src/models/characterModel.ts (hunk 10 above)
Comment (luke-dowling) [reply server/src/models/characterModel.ts]: I think that was the only one thats missing for now, it depends if the function being called inside another try catch block or not, only one top level is needed, then we can just throw errors to the top level trycatch -> any async await we need one though just incase

File: server/src/routes/characterRoutes.ts (hunk 11 above)
Comment (luke-dowling) [reply server/src/routes/characterRoutes.ts:14]: Yes that is actually a really good idea, i think for this one line it cuts down on code but if we are planning more here maybe it would be better, for now lets leave it like this

File: server/src/util/crudOps.ts [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+    try {
+      const data = await Model.create(req.body)
+
+      res.status(201).json({
+        status: 'success',
+        data,
+      })
+    } catch (error) {
+      console.error(error)
+      return next()
```
Comment (luke-dowling) [inline server/src/util/crudOps.ts:18]: we should put the error in the next: next(error)

**Review by luke-dowling — APPROVED** [review]
This looks good, lets merge it in :mechanical_arm:
